src/psyclone/psyir/nodes/loop.py

```diff
@@ -578,3 +579,53 @@ def gen_code(self, parent):
             if not kernel.module_inline:
                 parent.add(UseGen(parent, name=kernel._module_name, only=True,
                                   funcnames=[kernel._name]))
+
+    @property
+    def is_outermost(self):
+        '''
+        :returns: True if the loop has no ancestors, otherwise False
+        :rtype: bool
+        '''
+        return self.ancestor(Loop) is None
+
+    @property
+    def is_perfectly_nested(self):
+        '''
+        Determine whether the Loop nest of descendents of a Loop (inclusive) is
+        'perfect', i.e., each level excecpt the deepest contains only the next
+        Loop.
+
+        Note that we ignore Nodes of certain types.
+
+        :returns: True if the Loop nest is perfect, otherwise False
+        :rtype: bool
+        '''
+        loop_count = 0
+        next_loop = self
+        while isinstance(next_loop, Loop):
+            loop_count += 1
+
+            # If it has more than one child, the next loop will not be
+            # perfectly nested, so stop searching. If there is no child, we
+            # have an empty loop and we also stop here.
+            if len(next_loop.loop_body.children) != 1:
+                if (next_loop.loop_body.children and
+                        isinstance(next_loop.loop_body[0], Loop)):
+                    return False
+                return len(self.walk(Loop)) == loop_count
+
+            # If there are no more loops then check all the expected loops
+            # have been counted.
+            next_loop = next_loop.loop_body[0]
+            if not isinstance(next_loop, Loop):
+                return len(self.walk(Loop)) == loop_count
+        return True
+
+    @property
+    def nest_variable_names(self):
+        '''
+        :returns: Names of all variables within the Loop nest of descendents
+            of this Loop, inclusive.
+        :rtype: tuple[str]
```

Comment:
Just to emphasise, I think it would be more in-keeping with the PSyIR to return Symbols here. However, if you can show me how you use this then that will inform my opinion :-)